app/database/build_store.py

```diff
@@ -0,0 +1,307 @@
+        build._artifacts_zip_file = q_build.artifacts_zip_file
+        build._error_message = q_build.error_message
+        build._postbuild_tasks_are_finished = bool(int(q_build.postbuild_tasks_are_finished))
+        build.setup_failures = q_build.setup_failures
+        build._timing_file_path = q_build.timing_file_path
+
+        # Manually set the state machine timestamps
+        build._state_machine._transition_timestamps = {
+            BuildState.QUEUED: q_build.queued_ts,
+            BuildState.FINISHED: q_build.finished_ts,
+            BuildState.PREPARED: q_build.prepared_ts,
+            BuildState.PREPARING: q_build.preparing_ts,
+            BuildState.ERROR: q_build.error_ts,
+            BuildState.CANCELED: q_build.canceled_ts,
+            BuildState.BUILDING: q_build.building_ts
+        }
+        build._state_machine._fsm.current = BuildState[q_build.state]
+
+        build_artifact = BuildArtifact(q_build.build_artifact_dir)
+
+        directories = []
+        for directory in q_failed_artifact_directories:
+            directories.append(directory.failed_artifact_directory)
+        build_artifact._failed_artifact_directories = directories
+
+        pairs = []
+        for pair in q_failed_subjob_atom_pairs:
+            pairs.append((pair.subjob_id, pair.atom_id))
+        build_artifact._q_failed_subjob_atom_pairs = pairs
+
+        build._build_artifact = build_artifact
+
+        atoms_by_subjob_id = {}
+        for atom in q_build_atoms:
+            atoms_by_subjob_id.setdefault(atom.subjob_id, [])
+            atoms_by_subjob_id[atom.subjob_id].append(Atom(
+                atom.command_string,
+                atom.expected_time,
+                atom.actual_time,
+                atom.exit_code,
+                atom.state,
+                atom.atom_id,
+                atom.subjob_id
+            ))
+
+        subjobs = OrderedDict()
+        for subjob in q_build_subjobs:
+            atoms = atoms_by_subjob_id[subjob.subjob_id]
+            # Add atoms after subjob is created so we don't alter their state on initialization
+            subjob_to_add = Subjob(build_id, subjob.subjob_id, build.project_type, job_config, [])
+            subjob_to_add._atoms = atoms
+            subjob_to_add.completed = subjob.completed
+            subjobs[subjob.subjob_id] = subjob_to_add
+        build._all_subjobs_by_id = subjobs
+
+        # Place subjobs into correct queues within the build
+        build._unstarted_subjobs = Queue(maxsize=len(subjobs))
+        build._finished_subjobs = Queue(maxsize=len(subjobs))
+        for _, subjob in subjobs.items():
+            if subjob.completed:
+                build._finished_subjobs.put(subjob)
+            else:
+                build._unstarted_subjobs.put(subjob)
+
+        # Spend preparation coin if this build is already completed
+        if q_build.completed:
+            build._preparation_coin.spend()
```

Comment:
I point this issue out in the "Points of Concern & Things Left to Do" section of this PR description, I think we need to redefine our goal here with what we consider a reconstructed build